Add Groq Integration, Environment Variable Improvements, and CLI Updates

0bserver07 opened the pull request:
This pull request introduces several changes and improvements:


Groq integration updates:

- Updated the tool usage to improve groq integration (commit: da21777).
- Added groq support to the project (commit: c4ae171).
- Updated groq to use a proper tool approach, similar to the OpenAI client (commit: 1eaee28).
- Added the rest of the groq models (commit: fb442b4).

Environment Variables:

- Switched to using environment variables for more portable and secure configuration (commit: 1357b1f).

CLI and other adjustments:
- Updated pyproc for the CLI (commit: 0ce2fd2).
- Improved CLI implementation so that MindAgents is now runnable as a CLI (commit: 9e43de1).

Additional minor updates:
- Updated .gitignore to add repomix-output.md (commit: 46b8b24).
- Please review the changes and let me know if any adjustments are needed.

0bserver07 committed:
making mindagents runnable as cli

0bserver07 committed:
pyproc update for the cli

0bserver07 committed:
added the rest of the groq models

0bserver07 committed:
updating groq to make use of proper tool use, similar to the openai client

0bserver07 committed:
added repomix-output.md to gitignore

0bserver07 committed:
added groq

0bserver07 committed:
using ENV variables might be more portable and secure

0bserver07 committed:
the tool use needed to be updated for groq integration

0bserver07 committed:
removed distil

BBischof commented on src/mind_agents/core/display.py:
```diff
@@ -1,13 +1,13 @@
 from rich.panel import Panel
 from rich.table import Table
 
+from mind_agents.core.game import GameState, GameStateInfo, PlayerAction
```
Usually bar imports come first, and all imports are alphabetical

0bserver07 replied:
this commit organizes the import, and I noticed some imports not being used, a reshuffle in that top part:
223a0c36d5f49c763389feefa6177f230c0bc8c2

Updated:
:/ i removed the console by mistake: 709d9aecd3d4cab3ec93922ed1880760b990a8bd

BBischof commented on README.md:
````diff
@@ -191,19 +192,19 @@ You can specify which model each player should use when running the game. To see
 ```bash
 # List all available models
-uv run python -m src.play --list-models
+mind_agents --list-models
````
I prefer to encourage UV in the instructions as I think it simplifies installation and makes it much faster

0bserver07 replied:
this commit brings back the uv commands, 1b728a4a69fa3f884651f58ace6fb044eb399905, the --active command is just in case the end-user have something like pyenv running.

BBischof commented on src/mind_agents/llm/types.py:
```diff
@@ -137,6 +137,63 @@ class ModelMetadata:
+    "QWEN_2.5_32B": ModelMetadata(
+        provider="groq",
+        model_id="qwen-2.5-32b",
+        context_length=32768,  # Placeholder value - adjust based on actual API spec
```
should clean up these comments probably

0bserver07 replied:
good catch, comments cleaned, next i will test why the groq models might be failing besides the qwen:
7305880ee0bb0d33534459a878683e1ca62ba1f9

0bserver07 replied:
added a sim run with default size for max_tokens: ff0e00a961465abf623afcffe3be8924e656a448

```

uv run --active python -m mind_agents.core.simulator -p 2 -o 3 -l 1 -r 3 -m QWEN_2.5_CODER_32B
Running simulations with parameters:
- Player cards: 2
- Other cards: 3
- Played cards: [1]
- Model: QWEN_2.5_CODER_32B
- Resolution: 3

Found 528 valid scenarios
Running simulations: 100%|██████████████████████████████████████████████████████████████████████████| 528/528 [48:55<00:00,  5.56s/it]

Results saved to simulation_results/simulation_p2_h3_played1_r3_qwen_2.5_coder_32b.csv

Simulation Results:
    player_cards  other_cards played_cards  wait_time          model_name
0         (4, 7)            3         (1,)        2.0  QWEN_2.5_CODER_32B
1        (4, 10)            3         (1,)        5.0  QWEN_2.5_CODER_32B
2        (4, 13)            3         (1,)        2.0  QWEN_2.5_CODER_32B
3        (4, 16)            3         (1,)        2.0  QWEN_2.5_CODER_32B
4        (4, 19)            3         (1,)        2.0  QWEN_2.5_CODER_32B
..           ...          ...          ...        ...                 ...
523     (91, 97)            3         (1,)        5.0  QWEN_2.5_CODER_32B
524    (91, 100)            3         (1,)        5.0  QWEN_2.5_CODER_32B
525     (94, 97)            3         (1,)        5.0  QWEN_2.5_CODER_32B
526    (94, 100)            3         (1,)        5.0  QWEN_2.5_CODER_32B
527    (97, 100)            3         (1,)        5.0  QWEN_2.5_CODER_32B

[528 rows x 5 columns]

```

BBischof commented on src/mind_agents/play.py:
```diff
@@ -36,17 +37,43 @@
+    }
+    
+    # Check if all keys are available in environment
+    all_env_keys_present = True
```
Wont this prevent users from running this if they only have some of the api keys setup?

0bserver07 replied:
wow that's not good, I need to stop programming past 11, fixed, check below for partial configs to be allows:
373f66004ad8de171432f03dd4c83b45deaad0f2

BBischof commented on src/mind_agents/llm/providers/groq.py:
```diff
@@ -0,0 +1,143 @@
+        if self.config.max_tokens is not None:
+            completion_params["max_tokens"] = self.config.max_tokens
+
+        # Add reasoning_effort if supported and specified
```
where do you set these?

0bserver07 replied:
So I initially copied this like from the openai.py file:

```
if self.config.max_tokens is not None:
            completion_params["max_tokens"] = self.config.max_tokens
```

Then I realized, the end-user would have to modify the LLMConfig file to make this work:
```
config = LLMConfig(
    api_key=api_key,
    model_name=model_name,
    max_tokens=1000  # Set it here
)
```

instead, I added a new config variable that would allow for setting max_tokens per provider, what do you think?

3e65f696b323cf9fb4dc4dbcf1ce6bd419b56efd
and readme updated here:
77a4348e30d2aec37cabd0cb7f4af3573641eca5

BBischof reviewed (commented):
Left some comments

BBischof commented:
Getting this error when testing with `DEEPSEEK_QWEN_32B`; may need to handle function calling more carefully.

```
──────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────╯


ERROR    Failed to get valid response for player 1:                                                                                                             
         Response: Response(content='```json                                                                                                                    
         {                                                                                                                                                      
           "name": "wait_for_n_seconds",                                                                                                                        
           "argum...')                                                                                                                                          
         Model: DEEPSEEK_QWEN_32B                                                                                                                               
         Game state: {'card_number': 15, 'num_players': 3, 'total_other_cards': 2, 'all_cards': [15], 'played_cards': []}                                       
ERROR    Failed to get action for player 1                                                                                                                      
Traceback (most recent call last):
  File "/Users/bryanbischof/dev_other/mindAgents/.venv/bin/mind_agents", line 8, in <module>
    sys.exit(cli_main())
             ^^^^^^^^^^
  File "/Users/bryanbischof/dev_other/mindAgents/src/mind_agents/play.py", line 617, in cli_main
    asyncio.run(main(verbose=args.verbose, models=args.models, max_turns=args.max_turns))
  File "/Users/bryanbischof/.pyenv/versions/3.11.6/lib/python3.11/asyncio/runners.py", line 190, in run
    return runner.run(main)
           ^^^^^^^^^^^^^^^^
  File "/Users/bryanbischof/.pyenv/versions/3.11.6/lib/python3.11/asyncio/runners.py", line 118, in run
    return self._loop.run_until_complete(task)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/Users/bryanbischof/.pyenv/versions/3.11.6/lib/python3.11/asyncio/base_events.py", line 653, in run_until_complete
    return future.result()
           ^^^^^^^^^^^^^^^
  File "/Users/bryanbischof/dev_other/mindAgents/src/mind_agents/play.py", line 495, in main
    await play_round(game, verbose)
  File "/Users/bryanbischof/dev_other/mindAgents/src/mind_agents/play.py", line 264, in play_round
    raise RuntimeError(f"Critical error: Failed to get action for player {player.id}")
RuntimeError: Critical error: Failed to get action for player 1
```

BBischof commented on pyproject.toml:
```diff
@@ -17,8 +17,12 @@ dependencies = [
     "pydantic>=2.0.0",
     "tiktoken>=0.5.0",
     "httpx>=0.24.0",
+    "groq>=0.4.0",
```
I think we need to add to requirements.txt too

0bserver07 replied:
solid, added:
3c38da6e7ae61198457a74e52a9d4abb1c2be9ab

BBischof commented:
Missing a model here;

```


ERROR    Failed to get valid response for player 1:                                                                                                             
         Response: Response(error='Error code: 404 - {'error': {'message': 'The model `llama-3.3-70b-8192` does not exist or you do not have access to it.',    
         'type': 'invalid_request_error', 'code': 'model_not_found'}}')                                                                                         
         Model: LLAMA_3.3_70B                                                                                                                                   
         Game state: {'card_number': 37, 'num_players': 3, 'total_other_cards': 2, 'all_cards': [37], 'played_cards': []}                                       
ERROR    Failed to get action for player 1                                                                                                                      
Traceback (most recent call last):
  File "/Users/bryanbischof/dev_other/mindAgents/.venv/bin/mind_agents", line 8, in <module>
    sys.exit(cli_main())
             ^^^^^^^^^^
  File "/Users/bryanbischof/dev_other/mindAgents/src/mind_agents/play.py", line 617, in cli_main
    asyncio.run(main(verbose=args.verbose, models=args.models, max_turns=args.max_turns))
  File "/Users/bryanbischof/.pyenv/versions/3.11.6/lib/python3.11/asyncio/runners.py", line 190, in run
    return runner.run(main)
           ^^^^^^^^^^^^^^^^
  File "/Users/bryanbischof/.pyenv/versions/3.11.6/lib/python3.11/asyncio/runners.py", line 118, in run
    return self._loop.run_until_complete(task)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/Users/bryanbischof/.pyenv/versions/3.11.6/lib/python3.11/asyncio/base_events.py", line 653, in run_until_complete
    return future.result()
           ^^^^^^^^^^^^^^^
  File "/Users/bryanbischof/dev_other/mindAgents/src/mind_agents/play.py", line 495, in main
    await play_round(game, verbose)
  File "/Users/bryanbischof/dev_other/mindAgents/src/mind_agents/play.py", line 264, in play_round
    raise RuntimeError(f"Critical error: Failed to get action for player {player.id}")
RuntimeError: Critical error: Failed to get action for player 1
```

BBischof commented:
@0bserver07 i ran all the new models and here's the output:

https://gist.github.com/BBischof/15786e750a2354f1bdf515a3f6289ed5

0bserver07 committed:
readme prefers uv

0bserver07 committed:
fixed the bar import order and removed unused imports for readabiliy

0bserver07 committed:
removed comments in types for groq models

0bserver07 committed:
fixed the all_env_keys_present mistake, partial config or any env variable is allow now

0bserver07 committed:
feat(llm): add configurable max_tokens support

- Add default_max_tokens to ModelMetadata
- Support max_tokens configuration via:
  1. Environment variables (PROVIDER_MAX_TOKENS)
  2. Config file (~/.config/llm_keys/config.json)
  3. Model-specific defaults
- Update config loading to handle max_tokens settings

0bserver07 committed:
readme update for the max_tokens feature

0bserver07 committed:
updated requirements groq>=0.4.0

0bserver07 committed:
added 1 simulation with qwen_2.5_coder_32b

0bserver07 committed:
uhoh i fixed the console thing being removed

0bserver07 committed:
added better error logs

0bserver07 committed:
all the groq models work, updated logs, deepseek 32 needs work on tools

0bserver07 committed:
added special parser for DEEPSEEK_QWEN_32B, tool_use seems to be have a diff response

0bserver07 commented:
- Added better logging: c8b6597198f573d3f07deb3087c0db894e32bfbe

- From logs, I saw that I had set the wrong model ids:

Tested runs with: b1f9cfc0ed2e52738f399aa4a6938b2084139b0d

- LLAMA_3.3_70B
- LLAMA_3.2_90B_VISION
- LLAMA_3.2_11B_VISION
- DEEPSEEK_LLAMA_70B
- LLAMA_3.3_70B
- GEMMA_9B
- MIXTRAL_8X7B


- QWEN_2.5_32B: da2a56f215efd9fb0c00ba898c25611c616d857a

0bserver07 commented:
> Left some comments

Updated PR:


# LLM Integration and Game System Improvements

## Major Changes
1. Enhanced Groq LLM Integration
   - Added special parser for DEEPSEEK_QWEN_32B model
   - Improved tool call handling with support for both single and multiple tool calls
   - Added markdown code block cleaning for model responses
   - Enhanced logging throughout the Groq provider

2. Model Configuration Updates
   - Updated context lengths for various models to 128000 tokens
   - Added new models:
     - QWEN_2.5_32B
     - DEEPSEEK_QWEN_32B
     - Updated configurations for MIXTRAL_8X7B and other models

3. Error Handling and Logging
   - Implemented comprehensive error logging in player actions
   - Added detailed state information in error messages
   - Improved debugging capabilities with structured JSON logging

4. Dependencies and Documentation
   - Updated requirements to use groq>=0.4.0
   - Added documentation for max_tokens feature
   - Updated README to prefer uv for package management

## Technical Details
- Fixed content handling for None responses in Groq provider
- Implemented JSON parsing safety checks for tool calls
- Added structured logging with debug level information
- Corrected player model reference in error messages

## Testing Notes
- All Groq models are now functional
- Enhanced error reporting should help with debugging failed responses

## Ideas:
- running simulation for the groq models maybe.
- adding more comprehensive testing for new model configurations
- maybe making groq base url swapable to ollama.
